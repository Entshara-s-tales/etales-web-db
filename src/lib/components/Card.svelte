<script lang="ts">
	import type { Unit } from "../db/units";
	import { slide } from "svelte/transition";
	import { quintOut } from "svelte/easing";
	import Button from "./Button.svelte";

	export let unit: Unit;
	$: table = false;
	function openTable() {
		table = !table;
	}
</script>

<!-- href={`/units/${unit.id}`} -->

<!-- NOTE: This should not be necessary, seems like a bug in SvelteKit trying to render elements on server even without them being used -->
{#if unit}
	<div class="grid grid-cols-min shadow p-2 m-4">
		<a class="grid grid-cols-2" href="/#">
			<div>
				<h3 class="text-black">{unit.name}</h3>
				<span class="">Level: {unit.level}</span>
			</div>
			<img src="https://placekitten.com/40/40" alt="" />
		</a>
		<div class="">
			<Button><span on:click={openTable}>Drop table</span></Button>
			{#if table}
				<div transition:slide={{ delay: 100, duration: 300, easing: quintOut }}>
					{#each unit.dropTable as itemDrop}
						<p>Item: {itemDrop.item.name}</p>
						<p>Type: {itemDrop.item.type}</p>
						<p>State: {itemDrop.item.state}</p>
						<p>DropRate: {itemDrop.dropRate * 100}</p>
					{/each}
				</div>
			{/if}
		</div>
	</div>
{/if}
