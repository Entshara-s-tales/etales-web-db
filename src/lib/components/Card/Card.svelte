<script lang="ts">
	import { slide } from 'svelte/transition';
	import { quintOut } from 'svelte/easing';
	import type { Unit } from '../../db/units';
	import Button from '../Button/Button.svelte';

	import DropTable from '../DropTable/DropTable.svelte';

	export let unit: Unit;
	$: table = false;
	function openTable() {
		table = !table;
	}
</script>

<!-- href={`/units/${unit.id}`} -->

<!-- NOTE: This should not be necessary, seems like a bug in SvelteKit trying to render elements on server even without them being used -->
{#if unit}
	<div class="grid grid-cols-min shadow p-2 m-4 relative">
		<a class="grid grid-cols-2 " href="/#">
			<div>
				<h3 class="">{unit.name}</h3>
				<span class="">Level: {unit.level}</span>
			</div>
			<img src="https://placekitten.com/256/256" alt="" />
		</a>
		<div class="relative">
			<div class="mt-4 self-end place-self-center text-center">
				<Button onClick={openTable}>Drop table</Button>
			</div>
		</div>
		{#if table}
			<div
				class="z-10 w-full absolute left-0 top-0 shadow bg-white"
				transition:slide={{ delay: 100, duration: 300, easing: quintOut }}
			>
				<DropTable onClose={openTable} {unit} />
			</div>
		{/if}
	</div>
{/if}
