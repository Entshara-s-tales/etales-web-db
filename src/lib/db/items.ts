/**
 * Unknown right now, probably should be some kind of an enum
 * Seems that there's dress, head, unkown/not defined, weapon/wings, Miscelaneous
 */
export type ItemType = string;

// This isn't probably right though.
export type ItemState = 'passive';

export interface Item {
	name: string;
	type: ItemType;
	state: ItemState;
	description?: string;
}

export const MockItems: Item[] = [
	{
		name: 'Potato dress',
		type: 'Dress',
		state: 'passive',
		description: 'This item has a very calm aura',
	},
	{
		name: 'cat',
		type: 'Miscelaneous',
		state: 'passive',
	},
	{
		name: 'dog',
		type: 'Miscelaneous',
		state: 'passive',
		description: 'It definitely is not a cat!',
	},
];
