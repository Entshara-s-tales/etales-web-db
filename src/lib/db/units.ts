import { MockItems } from "./items";
import type { Item } from "./items";

export interface DropTable {
	item: Item;
	/**
	 * As percentages, normalized by 100 in frontend
	 */
	dropRate: number;
}

export interface Unit {
	name: string;
	id: number;
	// n <= 0 && n <= 99
	level: number;
	dropTable: DropTable[];
}

const MockDropTable = [
	{
		item: MockItems[0],
		dropRate: 0.1,
	},
	{
		item: MockItems[1],
		dropRate: 0.001,
	},
	{
		item: MockItems[2],
		dropRate: 0.8,
	},
];

export const MockUnits: Unit[] = [
	{
		name: "Spider pig",
		id: 0,
		level: 54,
		dropTable: MockDropTable,
	},
	{
		name: "Baby Bear",
		id: 64,
		level: 2,
		dropTable: MockDropTable,
	},
	{
		name: "Murloc Tiderunner",
		id: 93,
		level: 8,
		dropTable: MockDropTable,
	},
	{
		name: "Murloc Tiderunner Murloc Tiderunner",
		id: 93,
		level: 8,
		dropTable: MockDropTable,
	},
	{
		name: "Murloc Tiderunner Murloc Tiderunner Murloc Tiderunner",
		id: 93,
		level: 8,
		dropTable: MockDropTable,
	},
];
