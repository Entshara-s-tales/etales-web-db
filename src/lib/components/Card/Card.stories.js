import { MockUnits } from '$lib/db/units';
import Card from './Card.svelte';

export default {
  title: 'lib/components/Card',
  component: Card,
};

const Template = ({ onClick, ...args }) => ({
  Component: Card,
  props: args,
});

export const SampleUsage = Template.bind({});
SampleUsage.args = {
  label: 'Card',
  unit: MockUnits[0]
};
