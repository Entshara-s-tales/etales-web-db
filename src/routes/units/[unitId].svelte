<script lang="ts">
	import { MockUnits } from '$lib/db/units';

	export let unitId: number;
	const unit = MockUnits.find((x) => x.id === unitId);
</script>

{#if unit}
	<div class="flex flex-row">
		<h3 class="flex-grow text-black">{unit.name}</h3>
		<img src="https://placekitten.com/120/120" alt="" />
	</div>
{/if}
