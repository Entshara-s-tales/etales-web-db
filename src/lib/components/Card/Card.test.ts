import { MockUnits } from '$lib/db/units';
import { render } from '@testing-library/svelte';
import Card from './Card.svelte';

describe('<Card/>', () => {
	it('matches snapshot', () => {
		const results = render(Card, {
			unit: MockUnits[0],
		});
		expect(results.container).toMatchSnapshot();
	});
});
