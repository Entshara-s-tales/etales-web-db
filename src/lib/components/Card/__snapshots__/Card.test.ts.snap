// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card/> matches snapshot 1`] = `
<body>
  <div>
    <div
      class="grid grid-cols-min shadow p-2 m-4"
    >
      <a
        class="grid grid-cols-2"
        href="/#"
      >
        <div>
          <h3
            class="text-black"
          >
            Spider pig
          </h3>
           
          <span
            class=""
          >
            Level: 
            54
          </span>
        </div>
         
        <img
          alt=""
          src="https://placekitten.com/40/40"
        />
      </a>
       
      <div
        class="relative"
      >
        <button
          class="text-black rounded"
          type="button"
        >
          <span>
            Drop table
          </span>
        </button>
         
      </div>
    </div>
    
  </div>
</body>
`;
